Implement AWS Bedrock provider for AI Gateway (#9598)
mlflow/mlflow: docs/source/gateway/index.rst

Reviewer: can we add a minimum example in `examples/gateway`?

below can be used as a helpful guide when configuring a given route for any newly released model types as they become available with a given provider.
``N/A`` means that the provider currently doesn't support the route type.

+--------------------+--------------------------+--------------------+------------------+-----------------------------+--------------------------+--------------------------+--------------------------+

> | Route Type         | OpenAI                   | MosaicML           | Anthropic        | Cohere                      | Azure OpenAI             | MLflow                   | AWS Bedrock              |

Reviewer 2: Thanks so much for adding this to docs! Can we add an example in https://github.com/mlflow/mlflow/tree/master/examples/gateway as well? This is awesome!

| **anthropic_api_key**   | Yes      | N/A                      | This is the API key for the Anthropic service.        |
+-------------------------+----------+--------------------------+-------------------------------------------------------+

AWS Bedrock
+++++++++++

> Top-level model configuration for AWS Bedrock routes must be one of the two supported two authentication modes: key-based and role-based.

Reviewer 3: ```suggestion
Top-level model configuration for AWS Bedrock routes must be one of the following two supported authentication modes: `key-based` or `role-based`.
```

+--------------------------+----------+------------------------------+-------------------------------------------------------+
| Configuration Parameter  | Required | Default                      | Description                                           |
+==========================+==========+==============================+=======================================================+
| **aws_config**           | No       |                              | An object with either the key-based or role-based     |
|                          |          |                              | schema below.                                         |
+--------------------------+----------+------------------------------+-------------------------------------------------------+

> To use key-based authentication, define an AWS Bedrock route with usual ``aws_*`` settings.  For testing/development purposes these can be long-lived keys to a and IAM User or short-lived keys to an IAM role.  For long-term use, these *must* be long-lived keys to an IAM user authorized to use Bedrock.

Reviewer 3: ```suggestion
To use key-based authentication, define an AWS Bedrock route with the required fields below. 

.. note::
  
  If using a configured route purely for development or testing, utilizing an IAM User role or a temporary short-lived standard IAM role are recommended; while for production deployments, a standard long-expiry IAM role is recommended to ensure that the route is capable of handling authentication for a long period. If the authentication expires and a new set of keys need to be supplied, the route must be recreated in order to persist the new keys.
```

+--------------------------+----------+------------------------------+-------------------------------------------------------+
| **aws_secret_access_key**| Yes      |                              | AWS secret access key for the IAM user/role           |
|                          |          |                              | authorized to use bedrock                             |
+--------------------------+----------+------------------------------+-------------------------------------------------------+
| **aws_access_key_id**    | Yes      |                              | AWS access key ID for the IAM user/role               |
|                          |          |                              | authorized to use Bedrock                             |
+--------------------------+----------+------------------------------+-------------------------------------------------------+
| **aws_session_token**    | No       | None                         | Optional session token, if required                   |
+--------------------------+----------+------------------------------+-------------------------------------------------------+

> Alternatively, for role-based authentication, define AWS Bedrock route with an a IAM Role authorized to access Bedrock.  AI Gateway will attempt to assume this role with using the standard credential provider chain, and will renew the role credentials if they're expired.

Reviewer 3: ```suggestion
Alternatively, for role-based authentication, an AWS Bedrock route can be defined and initialized with an a IAM Role  ARN that is authorized to access Bedrock.  The MLflow AI Gateway will attempt to assume this role with using the standard credential provider chain and will renew the role credentials if they have expired.
```
